tools/pyvmi/libvmi_cdef.h

```diff
@@ -0,0 +1,339 @@
+
+    VMI_CONFIG_STRING,            /**< config string provided */
+
+    VMI_CONFIG_GHASHTABLE,        /**< config GHashTable provided */
+} vmi_config_t;
+
+// vmi_mode
+typedef enum vmi_mode {
+
+    VMI_XEN, /**< libvmi is monitoring a Xen VM */
+
+    VMI_KVM, /**< libvmi is monitoring a KVM VM */
+
+    VMI_FILE, /**< libvmi is viewing a file on disk */
+} vmi_mode_t;
+
+// vmi_init_error_t
+typedef enum vmi_init_error {
+
+    VMI_INIT_ERROR_NONE, /**< No error */
+
+    VMI_INIT_ERROR_DRIVER_NOT_DETECTED, /**< Failed to auto-detect hypervisor */
+
+    VMI_INIT_ERROR_DRIVER, /**< Failed to initialize hypervisor-driver */
+
+    VMI_INIT_ERROR_VM_NOT_FOUND, /**< Failed to find the specified VM */
+
+    VMI_INIT_ERROR_PAGING, /**< Failed to determine or initialize paging functions */
+
+    VMI_INIT_ERROR_OS, /**< Failed to determine or initialize OS functions */
+
+    VMI_INIT_ERROR_EVENTS, /**< Failed to initialize events */
+
+    VMI_INIT_ERROR_SHM, /**< Failed to initialize SHM */
+
+    VMI_INIT_ERROR_NO_CONFIG, /**< No configuration was found for OS initialization */
+
+    VMI_INIT_ERROR_NO_CONFIG_ENTRY, /**< Configuration contained no valid entry for VM */
+} vmi_init_error_t;
+
+// os_t
+typedef enum os {
+
+    VMI_OS_UNKNOWN,  /**< OS type is unknown */
+
+    VMI_OS_LINUX,    /**< OS type is Linux */
+
+    VMI_OS_WINDOWS   /**< OS type is Windows */
+} os_t;
+
+typedef enum translation_mechanism {
+    VMI_TM_INVALID,         /**< Invalid translation mechanism */
+    VMI_TM_NONE,            /**< No translation is required, address is physical address */
+    VMI_TM_PROCESS_DTB,     /**< Translate addr via specified directory table base. */
+    VMI_TM_PROCESS_PID,     /**< Translate addr by finding process first to use its DTB. */
+    VMI_TM_KERNEL_SYMBOL    /**< Find virtual address of kernel symbol and translate it via kernel DTB. */
+} translation_mechanism_t;
+
+// vmi_arch_t
+typedef enum arch {
+    VMI_ARCH_UNKNOWN,        /**< Unknown architecture */
+    VMI_ARCH_X86,            /**< x86 32-bit architecture */
+    VMI_ARCH_X86_64,         /**< x86 64-bit architecture */
+    VMI_ARCH_ARM32,          /**< ARM 32-bit architecture */
+    VMI_ARCH_ARM64           /**< ARM 64-bit architecture */
+} vmi_arch_t;
+
+// page_size_t
+typedef enum page_size {
+
+    VMI_PS_UNKNOWN  = 0,         /**< page size unknown */
+
+    VMI_PS_1KB      = 0x400,     /**< 1KB */
+
+    VMI_PS_4KB      = 0x1000,    /**< 4KB */
+
+    VMI_PS_16KB     = 0x4000,    /**< 16KB */
+
+    VMI_PS_64KB     = 0x10000,   /**< 64KB */
+
+    VMI_PS_1MB      = 0x100000,  /**< 1MB */
+
+    VMI_PS_2MB      = 0x200000,  /**< 2MB */
+
+    VMI_PS_4MB      = 0x400000,  /**< 4MB */
+
+    VMI_PS_16MB     = 0x1000000, /**< 16MB */
+
+    VMI_PS_32MB     = 0x2000000, /**< 32MB */
+
+    VMI_PS_512MB    = 0x2000000,  /**< 512MB */
+
+    VMI_PS_1GB      = 0x4000000,  /**< 1GB */
+
+} page_size_t;
+
+// page_info_t
+typedef struct page_info {
```

Comment:
In case you don't need direct access to the struct fields, you can replace them with a literal "...". The compiler will fill the rest of the information for you. This makes the code more portable as, if the struct gets changed, you just need to compile it against the new header.

See: http://cffi.readthedocs.io/en/latest/overview.html#abi-versus-api

Reply:
Done, thanks